JS value being parsed as a TS type argument

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!--
Please try to reproduce the issue with the latest published version. It may have already been fixed.
For npm: `typescript@next`
This is also the 'Nightly' version in the playground: http://www.typescriptlang.org/play/?ts=Nightly
-->
**TypeScript Version:**  3.8.0-dev.20200205

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** emscripten "type arguments"

**Code**

The [following code][1] (I added the declaration) was found in the distribution package for viz.js: https://github.com/mdaines/viz.js/releases/download/v2.1.2/lite.render.js.  This file is generated by emscripten, so the syntax is a little weird.

```ts
declare let [k, e, a, c, n]: number[];
if ((k | 0) < (e | 0) | (k | 0) > (a << 16 >> 16 | 0)) {
              c[n >> 2] = 1364;
}
```

[1]: https://www.typescriptlang.org/play/?ts=3.8.0-dev.20200205&ssl=1&ssc=1&pln=4&pc=2#code/CYUwxgNghgTiAEEQBd4G0DWAaeIdRzBwDsBdALnmIFcBbAIxBjVIG4AoASwDN4AKPhngAfeAAYAlPAA8-BKMkj+QhVIB8-KDNkBGAGzw1G-UslSA3u3jWbt22DTFDGgEyl4AXng6AzHoAsHAC+QA

**Expected behavior:**

This is valid JavaScript and should have the same parse as JS.

**Actual behavior:**

If this is parsed as TS, the part between the `<` and the `>` get pulled out as a type annotation, leaving

```
if ((k | 0)(a << 16 >> 16 | 0))
```

which is clearly incorrect, as it's trying to call a number.  If this is parsed as a JS file (with --allowJs), it produces an error:

```
$ tsc --allowJs a.js
a.js:2:14 - error TS8011: 'type arguments' can only be used in a .ts file.

2 if ((k | 0) < (e | 0) | (k | 0) > (a << 16 >> 16 | 0)) {
               ~~~~~~~~~~~~~~~~~~
```

Even if one makes the argument that this is not valid TS syntax, it should still parse the explicitly-JS input.

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior -->

https://www.typescriptlang.org/play/?ts=3.8.0-dev.20200205&ssl=1&ssc=1&pln=4&pc=2#code/CYUwxgNghgTiAEEQBd4G0DWAaeIdRzBwDsBdALnmIFcBbAIxBjVIG4AoASwDN4AKPhngAfeAAYAlPAA8-BKMkj+QhVIB8-KDNkBGAGzw1G-UslSA3u3jWbt22DTFDGgEyl4AXng6AzHoAsHAC+QA

**Related Issues:** <!-- Did you find other bugs that looked similar? -->

I was unable to find anything.

--- a/src/compiler/parser.ts
+++ b/src/compiler/parser.ts
@@ -4540,7 +4540,7 @@ namespace ts {
             }
 
             const tagName = parseJsxElementName();
-            const typeArguments = tryParseTypeArguments();
+            const typeArguments = (contextFlags & NodeFlags.JavaScriptFile) === 0 ? tryParseTypeArguments() : undefined;
             const attributes = parseJsxAttributes();
 
             let node: JsxOpeningLikeElement;
@@ -4797,7 +4797,8 @@ namespace ts {
                 const questionDotToken = parseOptionalToken(SyntaxKind.QuestionDotToken);
 
                 // handle 'foo<<T>()'
-                if (token() === SyntaxKind.LessThanToken || token() === SyntaxKind.LessThanLessThanToken) {
+                // parse template arguments only in TypeScript files (not in JavaScript files).
+                if ((contextFlags & NodeFlags.JavaScriptFile) === 0 && (token() === SyntaxKind.LessThanToken || token() === SyntaxKind.LessThanLessThanToken)) {
                     // See if this is the start of a generic invocation.  If so, consume it and
                     // keep checking for postfix expressions.  Otherwise, it's just a '<' that's
                     // part of an arithmetic expression.  Break out so we consume it higher in the
@@ -4854,6 +4855,11 @@ namespace ts {
         }
 
         function parseTypeArgumentsInExpression() {
+            if ((contextFlags & NodeFlags.JavaScriptFile) !== 0) {
+                // TypeArguments must not be parsed in JavaScript files to avoid ambiguity with binary operators.
+                return undefined;
+            }
+
             if (reScanLessThanToken() !== SyntaxKind.LessThanToken) {
                 return undefined;
             }

--- a/tests/baselines/reference/getJavaScriptSyntacticDiagnostics14.baseline
+++ b/tests/baselines/reference/getJavaScriptSyntacticDiagnostics14.baseline
@@ -1,8 +1,8 @@
 Syntactic Diagnostics for file '/tests/cases/fourslash/getJavaScriptSyntacticDiagnostics14.ts':
-/tests/cases/fourslash/a.js(1,5): error TS8011: Type arguments can only be used in TypeScript files.
+/tests/cases/fourslash/a.js(1,13): error TS1109: Expression expected.
 
 
 ==== /tests/cases/fourslash/a.js (1 errors) ====
     Foo<number>();
-        ~~~~~~
-!!! error TS8011: Type arguments can only be used in TypeScript files.
+                ~
+!!! error TS1109: Expression expected.

--- a/tests/baselines/reference/jsFileCompilationTypeArgumentSyntaxOfCall.errors.txt
+++ b/tests/baselines/reference/jsFileCompilationTypeArgumentSyntaxOfCall.errors.txt
@@ -1,19 +1,30 @@
-tests/cases/compiler/a.jsx(1,5): error TS8011: Type arguments can only be used in TypeScript files.
-tests/cases/compiler/a.jsx(2,5): error TS8011: Type arguments can only be used in TypeScript files.
-tests/cases/compiler/a.jsx(3,6): error TS8011: Type arguments can only be used in TypeScript files.
-tests/cases/compiler/a.jsx(4,6): error TS8011: Type arguments can only be used in TypeScript files.
+tests/cases/compiler/a.jsx(1,13): error TS1109: Expression expected.
+tests/cases/compiler/a.jsx(4,5): error TS1003: Identifier expected.
+tests/cases/compiler/a.jsx(4,14): error TS17002: Expected corresponding JSX closing tag for 'number'.
+tests/cases/compiler/a.jsx(4,20): error TS2657: JSX expressions must have one parent element.
+tests/cases/compiler/a.jsx(5,5): error TS1003: Identifier expected.
+tests/cases/compiler/a.jsx(5,6): error TS17008: JSX element 'number' has no corresponding closing tag.
+tests/cases/compiler/a.jsx(6,1): error TS1005: '</' expected.
 
 
-==== tests/cases/compiler/a.jsx (4 errors) ====
+==== tests/cases/compiler/a.jsx (7 errors) ====
     Foo<number>();
-        ~~~~~~
-!!! error TS8011: Type arguments can only be used in TypeScript files.
+                ~
+!!! error TS1109: Expression expected.
+    Foo<number>(1);
     Foo<number>``;
-        ~~~~~~
-!!! error TS8011: Type arguments can only be used in TypeScript files.
     <Foo<number>></Foo>;
-         ~~~~~~
-!!! error TS8011: Type arguments can only be used in TypeScript files.
+        ~
+!!! error TS1003: Identifier expected.
+                 ~~~~~~
+!!! error TS17002: Expected corresponding JSX closing tag for 'number'.
+                       ~
+!!! error TS2657: JSX expressions must have one parent element.
     <Foo<number>/>;
+        ~
+!!! error TS1003: Identifier expected.
          ~~~~~~
-!!! error TS8011: Type arguments can only be used in TypeScript files.
+!!! error TS17008: JSX element 'number' has no corresponding closing tag.
+    
+    
+!!! error TS1005: '</' expected.

--- a/tests/baselines/reference/jsFileCompilationTypeArgumentSyntaxOfCall.js
+++ b/tests/baselines/reference/jsFileCompilationTypeArgumentSyntaxOfCall.js
@@ -1,15 +1,15 @@
 //// [a.jsx]
 Foo<number>();
+Foo<number>(1);
 Foo<number>``;
 <Foo<number>></Foo>;
 <Foo<number>/>;
 
+
 //// [a.js]
-var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
-    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
-    return cooked;
-};
-Foo();
-Foo(__makeTemplateObject([""], [""]));
-<Foo></Foo>;
-<Foo />;
+Foo < number > ();
+Foo < number > (1);
+Foo < number > "";
+<Foo />, <number>></Foo>;
+<Foo />, <number>/>;
+</>;

--- a/tests/baselines/reference/jsxCheckJsxNoTypeArgumentsAllowed.errors.txt
+++ b/tests/baselines/reference/jsxCheckJsxNoTypeArgumentsAllowed.errors.txt
@@ -1,4 +1,8 @@
-tests/cases/conformance/jsx/file.jsx(4,17): error TS8011: Type arguments can only be used in TypeScript files.
+tests/cases/conformance/jsx/file.jsx(4,9): error TS2695: Left side of comma operator is unused and has no side effects.
+tests/cases/conformance/jsx/file.jsx(4,16): error TS1003: Identifier expected.
+tests/cases/conformance/jsx/file.jsx(4,17): error TS2693: 'Prop' only refers to a type, but is being used as a value here.
+tests/cases/conformance/jsx/file.jsx(4,17): error TS17008: JSX element 'Prop' has no corresponding closing tag.
+tests/cases/conformance/jsx/file.jsx(5,1): error TS1005: '</' expected.
 
 
 ==== tests/cases/conformance/jsx/component.d.ts (0 errors) ====
@@ -12,11 +16,19 @@ tests/cases/conformance/jsx/file.jsx(4,17): error TS8011: Type arguments can onl
         b: string
     }
     
-==== tests/cases/conformance/jsx/file.jsx (1 errors) ====
+==== tests/cases/conformance/jsx/file.jsx (5 errors) ====
     import { MyComp, Prop } from "./component";
     import * as React from "react";
     
     let x = <MyComp<Prop> a={10} b="hi" />; // error, no type arguments in js
+            ~~~~~~~
+!!! error TS2695: Left side of comma operator is unused and has no side effects.
+                   ~
+!!! error TS1003: Identifier expected.
                     ~~~~
-!!! error TS8011: Type arguments can only be used in TypeScript files.
+!!! error TS2693: 'Prop' only refers to a type, but is being used as a value here.
+                    ~~~~
+!!! error TS17008: JSX element 'Prop' has no corresponding closing tag.
     
+    
+!!! error TS1005: '</' expected.

--- a/tests/baselines/reference/jsxCheckJsxNoTypeArgumentsAllowed.js
+++ b/tests/baselines/reference/jsxCheckJsxNoTypeArgumentsAllowed.js
@@ -23,4 +23,5 @@ let x = <MyComp<Prop> a={10} b="hi" />; // error, no type arguments in js
 exports.__esModule = true;
 var component_1 = require("./component");
 var React = require("react");
-var x = <component_1.MyComp a={10} b="hi"/>; // error, no type arguments in js
+var x = <component_1.MyComp />, <Prop> a={10} b="hi" />; // error, no type arguments in js
+</>;

--- a/tests/baselines/reference/jsxCheckJsxNoTypeArgumentsAllowed.symbols
+++ b/tests/baselines/reference/jsxCheckJsxNoTypeArgumentsAllowed.symbols
@@ -36,7 +36,4 @@ import * as React from "react";
 let x = <MyComp<Prop> a={10} b="hi" />; // error, no type arguments in js
 >x : Symbol(x, Decl(file.jsx, 3, 3))
 >MyComp : Symbol(MyComp, Decl(file.jsx, 0, 8))
->Prop : Symbol(Prop, Decl(file.jsx, 0, 16))
->a : Symbol(a, Decl(file.jsx, 3, 21))
->b : Symbol(b, Decl(file.jsx, 3, 28))
 

--- a/tests/baselines/reference/jsxCheckJsxNoTypeArgumentsAllowed.types
+++ b/tests/baselines/reference/jsxCheckJsxNoTypeArgumentsAllowed.types
@@ -30,9 +30,12 @@ import * as React from "react";
 
 let x = <MyComp<Prop> a={10} b="hi" />; // error, no type arguments in js
 >x : JSX.Element
-><MyComp<Prop> a={10} b="hi" /> : JSX.Element
+><MyComp<Prop> a={10} b="hi" />; // error, no type arguments in js : JSX.Element
+><MyComp : JSX.Element
 >MyComp : typeof MyComp
->a : number
+><Prop> a={10} b="hi" />; // error, no type arguments in js : JSX.Element
+>Prop : any
 >10 : 10
->b : string
+
+> : any
 

--- a/tests/cases/compiler/jsFileCompilationTypeArgumentSyntaxOfCall.ts
+++ b/tests/cases/compiler/jsFileCompilationTypeArgumentSyntaxOfCall.ts
@@ -2,6 +2,7 @@
 // @noTypesAndSymbols: true
 // @filename: a.jsx
 Foo<number>();
+Foo<number>(1);
 Foo<number>``;
 <Foo<number>></Foo>;
 <Foo<number>/>;

